### ISSUE TITLE
QtWebEngine randomly freezes (no pattern detected yet)

### ISSUE THREAD

**Issue by ImmanuelHaffner** [qutebrowser/qutebrowser#4809]
**Version info (see `:version`)**:
qutebrowser v1.6.2
Git commit: 
Backend: QtWebEngine (Chromium 69.0.3497.128)

CPython: 3.7.3
Qt: 5.12.3
PyQt: 5.12.2

sip: 4.19.17
colorama: 0.4.1
pypeg2: 2.15
jinja2: 2.10.1
pygments: 2.4.0
yaml: 3.13
cssutils: no
attr: 19.1.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebKitWidgets: yes
pdf.js: 2.1.266 (/usr/share/pdf.js/build/pdf.js)
sqlite: 3.28.0
QtNetwork SSL: OpenSSL 1.1.1b  26 Feb 2019

Style: QFusionStyle
Platform: Linux-5.1.4-arch1-1-ARCH-x86_64-with-arch, 64bit
Linux distribution: Arch Linux (arch)
Frozen: False
Imported from /usr/lib/python3.7/site-packages/qutebrowser
Using Python from /usr/bin/python3
Qt library executable path: /usr/lib/qt/libexec, data path: /usr/share/qt

**Does the bug happen if you start with `--temp-basedir`?**:
Since the error occurs non-deterministically, I cannot tell.

**Description**
While browsing, qutebrowser rendering randomly stops. Resizing the window, switching the desktop, hiding and then displaying the window again, any keypress etc does not resurrect the frozen application. I observed that when a freeze happens and I kill the application and restart it and the previously opened tabs are reloaded, it is likely to freeze again very soon. Properly closing the application and starting fresh usually avoids freezing again. Therefore, I suspect that some websites trigger the freeze more often than others. Still, I don't know what exactly causes it and I cannot point to one site that causes the freeze.

Below is an excerpt of the debug log gathered until a freeze happened:
```
10:16:04 DEBUG    keyboard   basekeyparser:_debug_log:91 Got key: 0x4b / modifiers: 0x2000000 / text: 'K' / dry_run True
10:16:04 DEBUG    modes      modeman:_handle_keypress:199 match: 2, forward_unbound_keys: auto, passthrough: False, is_non_alnum: False, dry_run: True --> filter: True (focused: <PyQt5.QtWidgets.QWidget object at 0x7f2ba417ea68>)
10:16:04 DEBUG    modes      modeman:_handle_keypress:172 got keypress in mode KeyMode.normal - delegating to <qutebrowser.keyinput.modeparsers.NormalKeyParser>
10:16:04 DEBUG    keyboard   basekeyparser:_debug_log:91 Got key: 0x4b / modifiers: 0x2000000 / text: 'K' / dry_run False
10:16:04 DEBUG    keyboard   basekeyparser:_debug_log:91 Definitive match for 'K'.
10:16:04 DEBUG    keyboard   basekeyparser:_debug_log:91 Clearing keystring (was: K).
10:16:04 DEBUG    commands   command:run:534 command called: tab-next
10:16:04 DEBUG    commands   command:run:549 Calling qutebrowser.browser.commands.CommandDispatcher.tab_next(<qutebrowser.browser.commands.CommandDispatcher>, 1)
10:16:04 DEBUG    misc       app:on_focus_object_changed:836 Focus object changed: <PyQt5.QtWidgets.QWidget object at 0x7f2ba417ef78>
10:16:04 DEBUG    modes      tabbedbrowser:on_current_changed:701 Current tab changed, focusing <qutebrowser.browser.webengine.webenginetab.WebEngineTab tab_id=0 url='https://www.google.de/search?q=gtk2+and+qt5+apps+freeze&spell=1&sa=X&ved=0ahUKEwiZ3sXkn7viAhUFLlAKH…'>
10:16:04 DEBUG    modes      tabbedbrowser:on_current_changed:709 Mode before tab change: normal (mode_on_change = normal)
10:16:04 DEBUG    modes      modeman:leave:307 Ignoring leave request for KeyMode.hint (reason tab changed) as we're in mode KeyMode.normal
10:16:04 DEBUG    modes      modeman:leave:307 Ignoring leave request for KeyMode.caret (reason tab changed) as we're in mode KeyMode.normal
10:16:04 DEBUG    modes      modeman:leave:307 Ignoring leave request for KeyMode.insert (reason tab changed) as we're in mode KeyMode.normal
10:16:04 DEBUG    modes      modeman:leave:307 Ignoring leave request for KeyMode.passthrough (reason tab changed) as we're in mode KeyMode.normal
10:16:04 DEBUG    modes      tabbedbrowser:on_current_changed:721 Mode after tab change: normal (mode_on_change = normal)
10:16:04 DEBUG    modes      modeman:_handle_keypress:199 match: 2, forward_unbound_keys: auto, passthrough: False, is_non_alnum: False, dry_run: False --> filter: True (focused: <PyQt5.QtWidgets.QWidget object at 0x7f2ba417ef78>)
[1]    24939 killed     qutebrowser -d -l debug
```
It was somehow "caused" by pressing <Shift+k> to switch tabs.

FYI: I use AwesomeWM as window manager, which is a TWM.

**How to reproduce**

I have no idea how to reproduce this. It just happens during my every day work.

**Comment by The-Compiler** [comment]
Are you on intel graphics, with `xf86-video-intel` installed?

**Comment by ImmanuelHaffner** [comment]
Wow, fast reply!
Yes, that is exactly the package I use. In particular, I am running on `Intel Corporation HD Graphics 530` (from `lspci`).

**Comment by The-Compiler** [comment]
It's a bug with the 5.x Linux kernel in that Intel driver - I had the same issue, and when it happens, I also see Chromium and mpv freezing (especially when trying to fullscreen mpv). 

I've since switched to the X11 modesetting driver which gets picked when removing `xf86-video-intel`, and this seems to run much better. See #4641 for details. I wish I could debug this somehow, but since nobody found a way to consistently reproduce it, no luck so far...

**Comment by ImmanuelHaffner** [comment]
Let me add a probably important observation: Similar freezes also occur in two other applications I use, the Spotify for Linux client `aur/spotify` and `nteract`.

**Comment by ImmanuelHaffner** [comment]
All right, I will do that. Great support from your site. I highly appreciate your good work. Keep it up :)

**Comment by The-Compiler** [comment]
Thanks for the nice words! :smile: Note that (at least on my system) some monitor names as reported by `xrandr` changed with the different driver, so I've had to adjust some scripts.

**Comment by ImmanuelHaffner** [comment]
Hey @The-Compiler ,
thanks for pointing me to the Intel driver. I found a solution for the xf86-video-intel driver.

First, I disabled 3D acceleration in xorg.conf:
```
Section "Device"
    Identifier      "Intel"
    Driver          "intel"
    Option          "TripleBuffer" "true"
    Option          "AccelMethod" "sna"
    Option          "TearFree" "true"
    Option          "DRI" "false"
EndSection
```
(The acceleration method did not affect the freeze and i stick with `sna` as I had good experience with that so far.)

Then I set some kernel parameters for the Intel HD Graphics in my boot loader settings (systemd-boot):
```
i915.alpha_support=1
i915.edp_vswing=2
i915.enable_fbc=1
i915.enable_guc=2
i915.fastboot=1
i915.enable_dc=1
```

I don't know which of those settings in particular did the trick. I suspect it's either disabling 3D acceleration or `alpha_support`. The thing is, that silently the previously named setting `i915.preliminary_hw_support` was renamed to `alpha_support`. Just by browsing the kernel log I discovered that the kernel reported the flag as unknown so I searched online and found the new name of the setting. Maybe this is what is necessary.

I hope this helps :)
Regards, Immanuel
